lib/service/service.go

```diff
@@ -2509,6 +2568,251 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
+			}
+		}
+
+		// Loop over each application and create a server.
+		var applications []*services.App
+		for _, app := range process.Config.Apps.Apps {
+			publicAddr, err := getPublicAddr(authClient, app)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			a := &services.App{
+				Name:               app.Name,
+				URI:                app.URI,
+				PublicAddr:         publicAddr,
+				StaticLabels:       app.StaticLabels,
+				DynamicLabels:      services.LabelsToV2(app.DynamicLabels),
+				InsecureSkipVerify: app.InsecureSkipVerify,
+			}
+			if app.Rewrite != nil {
+				a.Rewrite = &services.Rewrite{
+					Redirect: app.Rewrite.Redirect,
+				}
+			}
+
+			applications = append(applications, a)
+		}
+		server := &services.ServerV2{
+			Kind:    services.KindAppServer,
+			Version: services.V2,
+			Metadata: services.Metadata{
+				Namespace: defaults.Namespace,
+				Name:      process.Config.HostUUID,
+			},
+			Spec: services.ServerSpecV2{
+				Hostname: process.Config.Hostname,
+				Version:  teleport.Version,
+				Apps:     applications,
+			},
+		}
+
+		authorizer, err := auth.NewAuthorizer(conn.Client, conn.Client, conn.Client)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		tlsConfig, err := conn.ServerIdentity.TLSConfig(nil)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		appServer, err = app.New(process.ExitContext(), &app.Config{
+			DataDir:      process.Config.DataDir,
+			AuthClient:   conn.Client,
+			AccessPoint:  authClient,
+			Authorizer:   authorizer,
+			TLSConfig:    tlsConfig,
+			CipherSuites: process.Config.CipherSuites,
+			GetRotation:  process.getRotation,
+			Server:       server,
+			OnHeartbeat: func(err error) {
+				if err != nil {
+					process.BroadcastEvent(Event{Name: TeleportDegradedEvent, Payload: teleport.ComponentApp})
+				} else {
+					process.BroadcastEvent(Event{Name: TeleportOKEvent, Payload: teleport.ComponentApp})
+				}
+			},
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Start the apps server. This starts the server, heartbeat (services.App),
+		// and (dynamic) label update.
+		appServer.Start()
+
+		// Create and start an agent pool.
+		agentPool, err = reversetunnel.NewAgentPool(process.ExitContext(),
+			reversetunnel.AgentPoolConfig{
+				Component:   teleport.ComponentApp,
+				HostUUID:    conn.ServerIdentity.ID.HostUUID,
+				ProxyAddr:   tunnelAddr,
+				Client:      conn.Client,
+				Server:      appServer,
+				AccessPoint: conn.Client,
+				HostSigner:  conn.ServerIdentity.KeySigner,
+				Cluster:     conn.ServerIdentity.Cert.Extensions[utils.CertExtensionAuthority],
+			})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		err = agentPool.Start()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Block and wait while the server and agent pool are running.
+		if err := appServer.Wait(); err != nil {
+			return trace.Wrap(err)
+		}
+		agentPool.Wait()
+
+		process.BroadcastEvent(Event{Name: AppsReady, Payload: nil})
```

Comment:
Also just noticed something: the two Wait calls are are blocking so these BroadcastEvent and Infof calls should probably come before them?